### PR TITLE
Added support for macros that allow to define when the connection is triggered

### PR DESCRIPTION
**Pull request by damianszczepanik**
With this change user can define additional conditions that must be met to trigger the notification. For example having generic build that builds many envs it is possible to get notification only for one (eg based on build params).

As quite a few plugins produces input for macro, powerful of this change is big and easy to extend and configure.

### CODE REVIEW

File: src/main/java/jenkins/plugins/office365connector/Macro.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+package jenkins.plugins.office365connector;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public class Macro {
+
+    /** Template that will be evaluated. */
+    private String template;
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Macro.java]: this is the template with the macro that is evaluated...

File: src/main/java/jenkins/plugins/office365connector/Macro.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    /** Expected value of the evaluated {@link #template}. */
+    private String value;
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Macro.java]: ... against this value

File: src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/config.jelly [hunk 3]
```diff
@@ -3,75 +3,98 @@
...
+
+						<f:entry title="Name" field="name">
+							<f:textbox name="name" value="${webhook.getName()}" />
+						</f:entry>
+
+						<f:entry title="URL" field="url">
+							<f:textbox name="url" value="${webhook.getUrl()}" />
+						</f:entry>
+
+						<f:advanced>
```
Comment (damianszczepanik) [inline src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/config.jelly]: Here we have advanced section which is by default collapsed (and user can expand it when need to apply modification). As this section becomes longer and longer (and I have plan to extend it in near future) it seems to be more user friendly that having everything expanded and scroll it all the time

File: src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/config.jelly [hunk 4]
```diff
@@ -3,75 +3,98 @@
...
+                                </f:entry>
+
+                                <f:entry title="Notify Repeated Failure">
+                                    <f:checkbox name="notifyRepeatedFailure" value="true" checked="${webhook.isNotifyRepeatedFailure()}"/>
+                                </f:entry>
+                            </f:section>
+
+                            <f:section title="Advanced configuration">
+                                <f:entry title="Macros" field="macros">
+                                    <f:repeatable name="macros" var="macro"
```
Comment (damianszczepanik) [inline src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/config.jelly]: Yes, user can define as many macros as wishes and if at least one is matched notification is sent.

File: src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<div align="help">
+  This section defines additional conditions that must be met to send the notification.
```
Comment (damianszczepanik) [inline src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html]: If this is not clear - let me know so I can provide better explanation...

File: src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+  It consists from pair of template and value. If template written
+  as <a href="https://wiki.jenkins.io/display/JENKINS/Token+Macro+Plugin">token macro</a>
```
Comment (damianszczepanik) [inline src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html]: ... but I believe this link and getting familiar with the plugin makes this feature understandable.

File: src/main/java/jenkins/plugins/office365connector/Macro.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package jenkins.plugins.office365connector;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public class Macro {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Macro.java]: I named it Macro but if you have better idea I'm happy to change

**Comment by damianszczepanik** [comment]
@outconn can you review and merge this change?

**Comment by damianszczepanik** [comment]
@outconn  @SrhebbarMicrosoft is there any chance for review and merge my PRs? Maybe you can consider assign to me push privileges as I'm [Jenkins member](https://github.com/orgs/jenkinsci/teams/cucumber-reports-plugin-developers/members) who has maintained own [plugin](https://github.com/jenkinsci/cucumber-reports-plugin)?

I;m going to add a few features that allow to customize messages etc.

**Comment by damianszczepanik** [comment]
Cool, any plan where this can be released?

File: src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html (hunk 5 above)
Comment (outconn) [reply src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html]: @damianszczepanik can you please provide examples for Macro, the notifications are not getting triggered for the ones we tried.

Comment (damianszczepanik) [reply src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html]: There are a few possibilities:
- add parameter to the build and compare this parameter to given value, eg `${param}` `ONE` so once you set `ONE` as the input for build execution, macro matches
- use predefined variable such as `${BUILD_NUMBER}` but this you need to update each time
- use any of the value that match together `abc` and the value `abc`

Comment (damianszczepanik) [reply src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html]: @outconn is there any wat to speed up merge progress? Some PR are waiting weeks to be reviewed and merged :(

Comment (outconn) [reply src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html]: @damianszczepanik  we are in the process of releasing the merged changes, Test pass is going on, but the Macro was not working so reached out to you. We will release the current set and take up other PRs. Sorry for the delay, we will be releasing the changes once a month from now on, will keep you updated n schedule.

@syamaraj123

Comment (casz) [reply src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html]: I wouldn't release now with the current state, I would add #42 before releasing You are going to break a lot of builds 👍

Comment (damianszczepanik) [reply src/main/resources/jenkins/plugins/office365connector/WebhookJobProperty/help-macros.html]: @outconn  I'm happy to have one release per month. Main problem which I see now is that PRs are waiting for review/feedback more than week and then it causes some delay with releasing, code rebase and stabilization.

**Pull request merged**